server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -320,4 +340,99 @@ public DirectoryReader getDirectoryReader() {
     }
+
+    /**
+     * Wraps an {@link IndexReader} with a cancellation Runnable task.
+     */
+    private static class CancellableIndexReader extends FilterDirectoryReader {
+
+        private final Cancellable checkCancelled;
+
+        private CancellableIndexReader(DirectoryReader in, Cancellable checkCancelled) throws IOException {
+            super(in, new SubReaderWrapper() {
+                @Override
+                public LeafReader wrap(LeafReader reader) {
+                    return new CancellableLeafReader(reader, checkCancelled);
+                }
+            });
+            this.checkCancelled = checkCancelled;
+        }
+
+        private void setCheckCancelled(Supplier<Runnable> checkCancelled) {
+            this.checkCancelled.setCancellable(checkCancelled);
+        }
+
+        @Override
+        protected DirectoryReader doWrapDirectoryReader(DirectoryReader in) {
+            return in;
+        }
+
+        @Override
+        public CacheHelper getReaderCacheHelper() {
+            return in.getReaderCacheHelper();
+        }
+    }
+
+    /**
+     * Wraps a leaf reader with a cancellable task
+     */
+    private static class CancellableLeafReader extends ExitableDirectoryReader.ExitableFilterAtomicReader {
+
+        private CancellableLeafReader(LeafReader leafReader, Cancellable checkCancelled)  {
+            super(leafReader, checkCancelled);
+        }
+
+        @Override
+        public NumericDocValues getNumericDocValues(String field) throws IOException {
+            return in.getNumericDocValues(field);
+        }
+
+        @Override
+        public BinaryDocValues getBinaryDocValues(String field) throws IOException {
+            return in.getBinaryDocValues(field);
+        }
+
+        @Override
+        public SortedDocValues getSortedDocValues(String field) throws IOException {
+            return in.getSortedDocValues(field);
+        }
+
+        @Override
+        public SortedNumericDocValues getSortedNumericDocValues(String field) throws IOException {
+            return in.getSortedNumericDocValues(field);
+        }
+
+        @Override
+        public SortedSetDocValues getSortedSetDocValues(String field) throws IOException {
+            return in.getSortedSetDocValues(field);
+        }
+    }
+
+    /**
+     * Implementation of {@link QueryTimeout} with a Runnable task.
+     */
+    private static class Cancellable implements QueryTimeout {
+
+        private Supplier<Runnable> cancellable;
+
+        public void setCancellable(Supplier<Runnable> cancellable) {
+            this.cancellable = cancellable;
+        }
+
+        @Override
+        public boolean shouldExit() {
+            assert cancellable != null : "checkCancelled must be set immediately after the construction of CancellableIndexReader";
+            if (cancellable.get() == null) {
+                return false;
+            }
+            cancellable.get().run();
+            return false;
+        }
```

Comment:
This implementation feels a bit awkward, I'd rather like to fork ExitableDirectoryReader entirely to not inherit from its QueryTimeout abstraction.

Reply:
I tried that in the 1st approach but this means we have to copy the whole `ExitablePointValues` to wrap the point values and therefore the `ExitableIntersectVisitor`.

Reply:
I don't mind copying it.

Reply:
+1 to copy